arc/src/ArcToken.sol

```diff
@@ -0,0 +1,765 @@
+    }
+
+    /**
+     * @dev Returns the URI for token metadata.
+     * @notice The URI should point to a JSON metadata object that follows the ERC-1155/OpenSea
+     * metadata standard format:
+     * {
+     *     "name": "Token Name",
+     *     "symbol": "SYMBOL",
+     *     "description": "Token description",
+     *     "image": "https://...", // URL to token image
+     *     "decimals": 18,
+     *     "properties": {
+     *         "assetName": "Asset Name",
+     *         "assetValuation": "1000000",
+     *         // other properties as needed
+     *     }
+     * }
+     */
+    function uri() public view returns (string memory) {
+        return _getArcTokenStorage().tokenURI;
+    }
+
+    // -------------- Minting and Burning --------------
+    /**
+     * @dev Mints new tokens to an account. Only accounts with MANAGER_ROLE can call this.
+     * The recipient must be whitelisted if transfers are restricted.
+     */
+    function mint(address to, uint256 amount) external onlyRole(MINTER_ROLE) {
+        _mint(to, amount);
+    }
+
+    /**
+     * @dev Burns tokens from an account, reducing the total supply. Only accounts with MANAGER_ROLE can call this.
+     */
+    function burn(address from, uint256 amount) external onlyRole(BURNER_ROLE) {
+        _burn(from, amount);
+    }
+
+    // -------------- Transfer Restrictions Toggle --------------
+    /**
+     * @dev Toggles transfer restrictions. When `transfersAllowed` is true, anyone can transfer tokens.
+     * When false, only whitelisted addresses can send/receive tokens.
+     */
+    function setTransfersAllowed(
+        bool allowed
+    ) external onlyRole(ADMIN_ROLE) {
+        _getArcTokenStorage().transfersAllowed = allowed;
+        emit TransfersRestrictionToggled(allowed);
+    }
+
+    //  Yield Distribution
+    /**
+     * @dev Sets or updates the ERC20 token to use for yield distribution (e.g., USDC).
+     * Only accounts with YIELD_MANAGER_ROLE can update this.
+     */
+    function setYieldToken(
+        address yieldTokenAddr
+    ) external onlyRole(YIELD_MANAGER_ROLE) {
+        if (yieldTokenAddr == address(0)) {
+            revert InvalidYieldTokenAddress();
+        }
+        _getArcTokenStorage().yieldToken = yieldTokenAddr;
+        emit YieldTokenUpdated(yieldTokenAddr);
+    }
+
+    /**
+     * @dev Get a preview of the yield distribution for token holders.
+     * This allows the yield distributor to check how much each holder would receive before
+     * actually distributing yield.
+     * @param amount The amount of yield token to preview distribution for
+     * @return holders Array of token holder addresses
+     * @return amounts Array of amounts each holder would receive
+     */
+    function previewYieldDistribution(
+        uint256 amount
+    ) external view returns (address[] memory holders, uint256[] memory amounts) {
+        ArcTokenStorage storage $ = _getArcTokenStorage();
```

Comment:
Could we take the common portions of this and `distributeYield` into a helper function so we know their output state is the same?